prompts/prompt_template_test.go

```diff
@@ -0,0 +1,412 @@
+	t.Parallel()
+
+	t.Run("Error", func(t *testing.T) {
+		t.Parallel()
+
+		pt, err := NewPromptTemplate(PromptTemplateOptions{
+			PromptTemplateBaseOptions: PromptTemplateBaseOptions{
+				InputVariables: []string{"testFormatPromptValue"},
+			},
+			Template:         "Hello, {{ .test }}",
+			ValidateTemplate: true,
+		})
+		require.NoError(t, err)
+
+		pt.opts.Template = "Hello, {{{ .testFormatPromptValue }}"
+		_, err = formatPromptValue(pt, map[string]any{
+			"testFormatPromptValue": "world",
+		})
+		require.Error(t, err)
+	})
+
+	t.Run("Success", func(t *testing.T) {
+		t.Parallel()
+
+		pt, err := NewPromptTemplate(PromptTemplateOptions{
+			PromptTemplateBaseOptions: PromptTemplateBaseOptions{
+				InputVariables: []string{"testFormatPromptValue"},
+			},
+			Template:         "Hello, {{ .testFormatPromptValue }}",
+			ValidateTemplate: true,
+		})
+		require.NoError(t, err)
+
+		spv, err := formatPromptValue(pt, map[string]any{
+			"testFormatPromptValue": "world",
+		})
+		require.NoError(t, err)
+		assert.Equal(t, "Hello, world", spv.String())
+	})
+}
+
+func TestMergePartialAndUserVariables(t *testing.T) {
+	t.Parallel()
+
+	t.Run("NilPartialVariables", func(t *testing.T) {
+		t.Parallel()
+
+		mergedVariables, err := mergePartialAndUserVariables(nil, map[string]any{
+			"test": "world",
+		})
+		require.NoError(t, err)
+		assert.Equal(t, map[string]any{
+			"test": "world",
+		}, mergedVariables)
+	})
+
+	t.Run("InvalidPartialVariableType", func(t *testing.T) {
+		t.Parallel()
+
+		_, err := mergePartialAndUserVariables(
+			map[string]any{
+				"test": 1234,
+			}, map[string]any{
+				"test": 1234,
+			},
+		)
+		require.Error(t, err)
+		assert.ErrorIs(t, err, ErrInvalidPartialVariableType)
+	})
+
+	t.Run("Success", func(t *testing.T) {
+		t.Parallel()
+
+		mergedVariables, err := mergePartialAndUserVariables(
+			map[string]interface{}{
+				"test":  "test",
+				"test2": "test2",
+				"test3": func() any {
+					return "test3"
+				},
+			}, map[string]any{
+				"test": "world",
+			},
+		)
+		require.NoError(t, err)
+		assert.Equal(t, mergedVariables, map[string]any{
+			"test":  "world",
+			"test2": "test2",
+			"test3": "test3",
+		})
+	})
+}
+
+func TestNewPromptTemplate(t *testing.T) {
+	t.Parallel()
+
+	_, err := NewPromptTemplate(PromptTemplateOptions{
+		PromptTemplateBaseOptions: PromptTemplateBaseOptions{
+			InputVariables: []string{"stop"},
+		},
+	})
+	require.Error(t, err)
+	assert.ErrorIs(t, err, ErrInputVariableReserved)
+
+	_, err = NewPromptTemplate(PromptTemplateOptions{
+		Template:         "Hello, {{ .test }}",
+		TemplateFormat:   "unknown",
+		ValidateTemplate: true,
+	})
+	require.Error(t, err)
+	assert.ErrorIs(t, err, ErrInvalidTemplateFormat)
+
+	_, err = NewPromptTemplate(PromptTemplateOptions{
+		Template:         "Hello, {{ .test }}",
+		ValidateTemplate: false,
+	})
+	require.NoError(t, err)
+}
+
+func TestPromptTemplateMergePartialAndUserVariables(t *testing.T) {
```

Comment:
Seems like a candidate for a table basted test

Reply:
Switched.